src/github.com/matrix-org/dendrite/mediaapi/writers/upload.go

```diff
@@ -0,0 +1,244 @@
+			}
+		}
+	}
+	return nil
+}
+
+// https://matrix.org/docs/spec/client_server/r0.2.0.html#post-matrix-media-r0-upload
+type uploadResponse struct {
+	ContentURI string `json:"content_uri"`
+}
+
+func removeDir(dir types.Path, logger *log.Entry) {
+	dirErr := os.RemoveAll(string(dir))
+	if dirErr != nil {
+		logger.Warnf("Failed to remove directory (%v): %q\n", dir, dirErr)
+	}
+}
+
+// Upload implements /upload
+//
+// This endpoint involves uploading potentially significant amounts of data to the homeserver.
+// This implementation supports a configurable maximum file size limit in bytes. If a user tries to upload more than this, they will receive an error that their upload is too large.
+// Uploaded files are processed piece-wise to avoid DoS attacks which would starve the server of memory.
+// TODO: Requests time out if they have not received any data within the configured timeout period.
+func Upload(req *http.Request, cfg *config.MediaAPI, db *storage.Database) util.JSONResponse {
+	logger := util.GetLogger(req.Context())
+
+	if req.Method != "POST" {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.Unknown("HTTP request method must be POST."),
+		}
+	}
+
+	// FIXME: This will require querying some other component/db but currently
+	// just accepts a user id for auth
+	userID, resErr := auth.VerifyAccessToken(req)
+	if resErr != nil {
+		return *resErr
+	}
+
+	r := &uploadRequest{
+		MediaMetadata: &types.MediaMetadata{
+			Origin:             cfg.ServerName,
+			ContentDisposition: types.ContentDisposition(req.Header.Get("Content-Disposition")),
+			ContentLength:      types.ContentLength(req.ContentLength),
+			ContentType:        types.ContentType(req.Header.Get("Content-Type")),
+			UploadName:         types.Filename(req.FormValue("filename")),
+			UserID:             types.MatrixUserID(userID),
+		},
+	}
+
+	if resErr = r.Validate(cfg.MaxFileSizeBytes); resErr != nil {
+		return *resErr
+	}
+
+	if len(r.MediaMetadata.UploadName) > 0 {
+		r.MediaMetadata.ContentDisposition = types.ContentDisposition(
+			"inline; filename*=utf-8''" + url.PathEscape(string(r.MediaMetadata.UploadName)),
+		)
+	}
+
+	logger.WithFields(log.Fields{
+		"Origin":              r.MediaMetadata.Origin,
+		"UploadName":          r.MediaMetadata.UploadName,
+		"Content-Length":      r.MediaMetadata.ContentLength,
+		"Content-Type":        r.MediaMetadata.ContentType,
+		"Content-Disposition": r.MediaMetadata.ContentDisposition,
+	}).Info("Uploading file")
+
+	writer, file, tmpDir, errorResponse := createTempFileWriter(cfg.BasePath, logger)
+	if errorResponse != nil {
+		return *errorResponse
+	}
+	defer file.Close()
+
+	// The limited reader restricts how many bytes are read from the body to the specified maximum bytes
+	// Note: the golang HTTP server closes the request body
+	limitedBody := io.LimitReader(req.Body, int64(cfg.MaxFileSizeBytes))
+	hasher := sha256.New()
+	reader := io.TeeReader(limitedBody, hasher)
```

Comment:
Why do we need this? I'm guessing to simultaneously calculate a streaming hash and dump to file? Some clarification would be good, we don't even say why we're calculating a hash!